components/expand-collapse/expand-collapse.js

```diff
@@ -0,0 +1,144 @@
+
+const states = {
+	PRECOLLAPSING: 'precollapsing', // setting up the styles so the collapse transition will run
+	COLLAPSING: 'collapsing', // in the process of collapsing
+	COLLAPSED: 'collapsed', // fully collapsed
+	PREEXPANDING: 'preexpanding', // setting up the styles so the expand transition will run
+	EXPANDING: 'expanding', // in the process of expanding
+	EXPANDED: 'expanded', // fully expanded
+};
+
+class ExpandCollapse extends LitElement {
+
+	static get properties() {
+		return {
+			expanded: { type: Boolean, reflect: true },
+			_height: { type: String },
+			_state: { type: String }
+		};
+	}
+
+	static get styles() {
+		return css`
+			:host {
+				display: block;
+			}
+
+			:host([hidden]) {
+				display: none;
+			}
+
+			.d2l-expand-collapse-container {
+				display: none;
+				overflow: hidden;
+				transition: height 400ms cubic-bezier(0, 0.7, 0.5, 1);
+			}
+
+			.d2l-expand-collapse-container:not([data-state="collapsed"]) {
+				display: block;
+			}
+
+			.d2l-expand-collapse-container[data-state="expanded"] {
+				overflow: visible;
+			}
+
+			/* prevent margin colapse on slotted children */
+			.d2l-expand-collapse-content:before,
+			.d2l-expand-collapse-content:after {
+				content: ' ';
+				display: table;
+			}
+
+			@media (prefers-reduced-motion: reduce) {
+				.d2l-expand-collapse-container {
+					transition: none;
+				}
+			}
+		`;
+	}
+
+	constructor() {
+		super();
+		this.expanded = false;
+		this._state = states.COLLAPSED;
+		this._height = '0';
+	}
+
+	updated(changedProperties) {
+		if (changedProperties.has('expanded')) {
+			this._expandedChanged(this.expanded);
+		}
+	}
+
+	render() {
+		const styles = { height: this._height };
+		return html`
+			<div class="d2l-expand-collapse-container" data-state=${this._state} @transitionend=${this._onTransitionEnd} style=${styleMap(styles)}>
+				<div class="d2l-expand-collapse-content">
+					<slot></slot>
+				</div>
+			</div>
+		`;
+	}
+
+	async _expandedChanged(val) {
+		if (val) {
+			if (reduceMotion) {
+				this._state = states.EXPANDED;
+				this._height = 'auto';
+			} else {
+				this._state = states.PREEXPANDING;
+				await this.updateComplete;
+				await new Promise((r) => requestAnimationFrame(() => requestAnimationFrame(r)));
+				if (this._state === states.PREEXPANDING) {
+					this._state = states.EXPANDING;
+					const content = this._getContent();
+					this._height = `${content.scrollHeight}px`;
+				}
+			}
+		} else {
+			if (reduceMotion) {
+				this._state = states.COLLAPSED;
+				this._height = '0';
+			} else {
+				this._state = states.PRECOLLAPSING;
+				const content = this._getContent();
+				this._height = `${content.scrollHeight}px`;
+				await this.updateComplete;
+				await new Promise((r) => requestAnimationFrame(() => requestAnimationFrame(r)));
+				if (this._state === states.PRECOLLAPSING) {
+					this._state = states.COLLAPSING;
+					this._height = '0';
+				}
+			}
+		}
```

Comment:
What do folks think about firing a `d2l-expand-collapse-state-change` (or similarly named) event when it completes its transition from expanded -> collapsed or vice versa? I can imagine consumers will want to know so that they can swap icons on openers, text, etc. For example if more/less used this it would need to switch its text and icon.

Reply:
I was thinking that could be done in the same callback that toggles with `expanded` attribute but I certainly don't think it'd hurt to add an event, there are probably some cases it would come in handy.

Reply:
Agreed.  I think an event could be useful.  An idea I've mulled over a little bit is to actually dispatch the event before the animation begins:
```
this.dispatchEvent(new CustomEvent(
    'd2l-expand', {
    bubbles: true,
    detail: { expandComplete: /* promise that resolves when transition complete */  }
    }
));

this.dispatchEvent(new CustomEvent(
    'd2l-collapse', {
    bubbles: true,
    detail: { collapseComplete: /* promise that resolves when transition complete */  }
    }
));
```
Or one event for both states could be fine too.  

It sorta solves the issue where a component dispatches and event, but not at the time when a consumer is interested, because it gives the consumer both.  I think there are a couple places were we dispatch an event at the beginning (ex. so that nest menu can translate at the same time as the dropdown height animates), but many times we are interested in when the animation is finished.

Reply:
^ the above mentioned problem is traditionally solved by dispatching different events for before/after.  The above approach is sorta cleaner, but not what consumers expect.

Reply:
That's a really good point, and I think consumers like more/less would absolutely want the event at the start. I like the pattern of including the promise in the detail.